modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,173 @@
+
+import org.tessellation.dag.domain.block.{BlockReference, DAGBlock}
+import org.tessellation.dag.l1.config.TipsConfig
+import org.tessellation.dag.l1.domain.block
+import org.tessellation.dag.l1.domain.block.{BlockService, Tips}
+import org.tessellation.ext.collection.MapRefUtils.MapRefOps
+import org.tessellation.security.Hashed
+import org.tessellation.security.hash.ProofsHash
+import org.tessellation.security.signature.Signed
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.PosInt
+import io.chrisdavenport.mapref.MapRef
+import io.estatico.newtype.ops._
+import monocle.macros.syntax.lens._
+
+object BlockService {
+
+  def make[F[_]: Sync](tipsConfig: TipsConfig): F[BlockService[F]] =
+    MapRef.ofConcurrentHashMap[F, ProofsHash, StoredBlock]().map(make(_, tipsConfig))
+
+  def make[F[_]: Sync](blocks: MapRef[F, ProofsHash, Option[StoredBlock]], tipsConfig: TipsConfig): BlockService[F] =
+    new BlockService[F] {
+
+      def areParentsAccepted(block: DAGBlock): F[Map[BlockReference, Boolean]] =
+        block.parent.traverse { ref =>
+          isBlockAccepted(ref).map(ref -> _)
+        }.map(_.toNem.toSortedMap)
+
+      def accept(hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        blocks(hashedBlock.proofsHash).modify {
+          case Some(storedBlock) =>
+            storedBlock match {
+              case WaitingBlock(_, usages) =>
+                (AcceptedBlock(hashedBlock, Accepted, usages, isTip = false).some, ())
+              case accepted: AcceptedBlock => (accepted.some, ())
+            }
+          case None => (AcceptedBlock(hashedBlock, Accepted, Set.empty, isTip = false).some, ())
+        }
+
+      def store(hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        blocks(hashedBlock.proofsHash).modify {
+          case None  => (WaitingBlock(hashedBlock.signed, Set.empty[ProofsHash]).some, ())
+          case other => (other, ())
+        }
+
+      def getWaiting: F[Map[ProofsHash, Signed[DAGBlock]]] =
+        blocks.toMap.map(_.collect { case (hash, WaitingBlock(block, _)) => hash -> block })
+
+      def handleTipsUpdate(hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        for {
+          allBlocks <- blocks.toMap
+          allTips = allBlocks.collect { case (hash, ab @ AcceptedBlock(_, _, _, true)) => hash -> ab }
+          parentHashToTip = hashedBlock.signed.value.parent.map(ref => ref.hash -> allTips.get(ref.hash)).toList.toMap
+          tipParents = parentHashToTip.collect { case (hash, Some(accepted)) => hash -> accepted }
+          notTipParents = (parentHashToTip -- tipParents.keySet).keySet
+          tipParentsSize = tipParents.size
+          tipParentsToRemove = tipParents.filter {
+            case (_, block) =>
+              allTips.size >= tipsConfig.minimumTipsCount + tipParentsSize &&
+                (
+                  allTips.size > tipsConfig.maximumTipsCount + tipParentsSize ||
+                    block.usages.size >= tipsConfig.maximumTipUsages - 1
```

Comment:
why `-1`? can you assign these conditions to some expressive names, like:
```scala
val aboveOrEqualRequiredMinimum = ...
val aboveRequiredMinimum
val areParentsUsedEnoughTimes
```
or sth like that

Reply:
`-1` because after adding a usage the `maximumTipUsages` count will be reached. Yeah I will try to extract conditions to properly named vals.